File: apps/mobile/ios/Happy/ViewModels/FriendsViewModel.swift
//
//  FriendsViewModel.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import Foundation
import Combine

/// Protocol for the friends API operations, enabling dependency injection and testing.
protocol FriendsAPIServiceProtocol {
    func fetchFriends() async throws -> [Friend]
    func fetchFriendRequests() async throws -> [FriendRequest]
    func sendFriendRequest(toUsername: String, message: String?) async throws -> FriendRequest
    func sendFriendRequestByQR(userId: String, publicKey: String) async throws -> FriendRequest
    func acceptFriendRequest(requestId: String) async throws -> Friend
    func declineFriendRequest(requestId: String) async throws
    func removeFriend(friendId: String) async throws
}

/// ViewModel for the friends list and social features.
///
/// Manages fetching, filtering, and real-time updates of the friend list,
/// pending friend requests, and friend request operations.
/// Uses `ObservableObject` for iOS 16 compatibility.
final class FriendsViewModel: ObservableObject {

    // MARK: - Published Properties

    /// All friends fetched from the server, sorted by online status then name.
    @Published private(set) var friends: [Friend] = []

    /// Pending friend requests (both incoming and outgoing).
    @Published private(set) var friendRequests: [FriendRequest] = []

    /// Whether a network request is in progress.
    @Published private(set) var isLoading: Bool = false

    /// Whether the initial load has completed.
    @Published private(set) var hasLoaded: Bool = false

    /// The current error message, if any.
    @Published var errorMessage: String?

    /// Whether to show the error alert.
    @Published var showError: Bool = false

    /// Text entered in the search bar.
    @Published var searchText: String = ""

    /// The current filter for the friends list.
    @Published var statusFilter: FriendStatusFilter = .all

    /// Whether a friend request is being sent.
    @Published private(set) var isSendingRequest: Bool = false

    /// Confirmation message after a successful action.
    @Published var confirmationMessage: String?

    /// Whether to show the confirmation alert.
    @Published var showConfirmation: Bool = false

    // MARK: - Computed Properties

    /// Friends filtered by the current search text and status filter.
    var filteredFriends: [Friend] {
        var result = friends

        // Apply status filter
        switch statusFilter {
        case .all:
            break
        case .online:
            result = result.filter { $0.status == .online || $0.status == .inSession }
        case .offline:
            result = result.filter { $0.status == .offline }
        }

        // Apply search filter
        if !searchText.isEmpty {
            result = result.filter {
                $0.displayName.localizedCaseInsensitiveContains(searchText) ||
                ($0.email?.localizedCaseInsensitiveContains(searchText) ?? false)
            }
        }

        return result
    }

    /// Count of online friends for badge display.
    var onlineFriendCount: Int {
        friends.filter { $0.isOnline || $0.status == .inSession }.count
    }

    /// Pending incoming friend requests.
    var incomingRequests: [FriendRequest] {
        friendRequests.filter { $0.status == .pending && $0.direction(for: currentUserId) == .received }
    }

    /// Pending outgoing friend requests.
    var outgoingRequests: [FriendRequest] {
        friendRequests.filter { $0.status == .pending && $0.direction(for: currentUserId) == .sent }
    }

    /// Count of pending incoming requests for badge display.
    var pendingRequestCount: Int {
        incomingRequests.count
    }

    /// Whether the friends list is empty after filtering.
    var isEmptyState: Bool {
        hasLoaded && filteredFriends.isEmpty
    }

    // MARK: - Dependencies

    private let apiService: FriendsAPIServiceProtocol
    private let syncService: SyncService
    private let currentUserId: String
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    /// Creates a new friends view model.
    ///
    /// - Parameters:
    ///   - apiService: The API service for friend operations. Defaults to the shared adapter.
    ///   - syncService: The sync service for real-time updates. Defaults to the shared instance.
    ///   - currentUserId: The current user's ID. Defaults to the stored account ID.
    init(
        apiService: FriendsAPIServiceProtocol? = nil,
        syncService: SyncService = .shared,
        currentUserId: String? = nil
    ) {
        self.apiService = apiService ?? FriendsAPIServiceAdapter()
        self.syncService = syncService
        self.currentUserId = currentUserId ?? KeychainHelper.readString(.accountId) ?? ""
    }

    // MARK: - Public Methods

    /// Fetches both the friend list and pending requests from the server.
    ///
    /// Sets `isLoading` during the request and updates the published properties on success.
    @MainActor
    func loadFriends() async {
        isLoading = true
        errorMessage = nil

        do {
            async let fetchedFriends = apiService.fetchFriends()
            async let fetchedRequests = apiService.fetchFriendRequests()

            let (friendsResult, requestsResult) = try await (fetchedFriends, fetchedRequests)
            friends = sortFriends(friendsResult)
            friendRequests = requestsResult.sorted { $0.createdAt > $1.createdAt }
            hasLoaded = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    /// Refreshes the friend list (pull-to-refresh).
    @MainActor
    func refresh() async {
        await loadFriends()
    }

    /// Sends a friend request to a user by username.
    ///
    /// - Parameters:
    ///   - username: The username to send the request to.
    ///   - message: An optional message to include with the request.
    @MainActor
    func sendFriendRequest(toUsername username: String, message: String? = nil) async {
        isSendingRequest = true
        errorMessage = nil

        do {
            let request = try await apiService.sendFriendRequest(toUsername: username, message: message)
            friendRequests.insert(request, at: 0)
            confirmationMessage = "Friend request sent to \(username)"
            showConfirmation = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isSendingRequest = false
    }

    /// Sends a friend request via QR code scan data.
    ///
    /// - Parameters:
    ///   - userId: The target user's ID from the QR code.
    ///   - publicKey: The target user's public key from the QR code.
    @MainActor
    func sendFriendRequestByQR(userId: String, publicKey: String) async {
        isSendingRequest = true
        errorMessage = nil

        do {
            let request = try await apiService.sendFriendRequestByQR(userId: userId, publicKey: publicKey)
            friendRequests.insert(request, at: 0)
            confirmationMessage = "Friend request sent!"
            showConfirmation = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isSendingRequest = false
    }

    /// Accepts a pending friend request.
    ///
    /// - Parameter request: The friend request to accept.
    @MainActor
    func acceptRequest(_ request: FriendRequest) async {
        do {
            let newFriend = try await apiService.acceptFriendRequest(requestId: request.id)
            // Add the new friend and remove the request
            friends = sortFriends(friends + [newFriend])
            friendRequests.removeAll { $0.id == request.id }
            confirmationMessage = "\(request.fromDisplayName) is now your friend"
            showConfirmation = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    /// Declines a pending friend request.
    ///
    /// - Parameter request: The friend request to decline.
    @MainActor
    func declineRequest(_ request: FriendRequest) async {
        do {
            try await apiService.declineFriendRequest(requestId: request.id)
            friendRequests.removeAll { $0.id == request.id }
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    /// Removes a friend from the friend list.
    ///
    /// - Parameter friend: The friend to remove.
    @MainActor
    func removeFriend(_ friend: Friend) async {
        do {
            try await apiService.removeFriend(friendId: friend.id)
            friends.removeAll { $0.id == friend.id }
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    /// Dismisses the current error.
    @MainActor
    func dismissError() {
        errorMessage = nil
        showError = false
    }

    /// Dismisses the confirmation message.
    @MainActor
    func dismissConfirmation() {
        confirmationMessage = nil
        showConfirmation = false
    }

    // MARK: - Private Methods

    /// Sorts friends by online status (online first), then by display name.
    private func sortFriends(_ list: [Friend]) -> [Friend] {
        list.sorted { lhs, rhs in
            let lhsPriority = statusSortPriority(lhs.status)
            let rhsPriority = statusSortPriority(rhs.status)
            if lhsPriority != rhsPriority {
                return lhsPriority < rhsPriority
            }
            return lhs.displayName.localizedCaseInsensitiveCompare(rhs.displayName) == .orderedAscending
        }
    }

    /// Returns a numeric priority for sorting (lower = higher priority).
    private func statusSortPriority(_ status: FriendStatus) -> Int {
        switch status {
        case .inSession: return 0
        case .online: return 1
        case .away: return 2
        case .offline: return 3
        }
    }
}

// MARK: - Friend Status Filter

/// Filter options for the friends list.
enum FriendStatusFilter: String, CaseIterable, Identifiable {
    case all = "All"
    case online = "Online"
    case offline = "Offline"

    var id: String { rawValue }
}

// MARK: - Friends API Adapter

/// Adapter that bridges `APIService` to `FriendsAPIServiceProtocol`.
///
/// Provides friend-specific endpoints using the existing actor-based `APIService`.
struct FriendsAPIServiceAdapter: FriendsAPIServiceProtocol {

    private let apiService: APIService

    init(apiService: APIService = .shared) {
        self.apiService = apiService
    }

    func fetchFriends() async throws -> [Friend] {
        struct FriendsResponse: Decodable {
            let friends: [Friend]
        }
        let response: FriendsResponse = try await apiService.fetch("/v1/friends")
        return response.friends
    }

    func fetchFriendRequests() async throws -> [FriendRequest] {
        struct RequestsResponse: Decodable {
            let requests: [FriendRequest]
        }
        let response: RequestsResponse = try await apiService.fetch("/v1/friends/requests")
        return response.requests
    }

    func sendFriendRequest(toUsername: String, message: String?) async throws -> FriendRequest {
        struct SendRequestBody: Encodable {
            let username: String
            let message: String?
        }
        let body = SendRequestBody(username: toUsername, message: message)
        return try await apiService.post("/v1/friends/requests", body: body)
    }

    func sendFriendRequestByQR(userId: String, publicKey: String) async throws -> FriendRequest {
        struct QRRequestBody: Encodable {
            let userId: String
            let publicKey: String
        }
        let body = QRRequestBody(userId: userId, publicKey: publicKey)
        return try await apiService.post("/v1/friends/requests/qr", body: body)
    }

    func acceptFriendRequest(requestId: String) async throws -> Friend {
        struct EmptyBody: Encodable {}
        return try await apiService.put("/v1/friends/requests/\(requestId)/accept", body: EmptyBody())
    }

    func declineFriendRequest(requestId: String) async throws {
        struct EmptyBody: Encodable {}
        struct EmptyResponse: Decodable {}
        let _: EmptyResponse = try await apiService.put("/v1/friends/requests/\(requestId)/decline", body: EmptyBody())
    }

    func removeFriend(friendId: String) async throws {
        struct EmptyBody: Encodable {}
        struct EmptyResponse: Decodable {}
        let _: EmptyResponse = try await apiService.put("/v1/friends/\(friendId)/remove", body: EmptyBody())
    }
}

File: apps/mobile/android/app/src/main/java/com/enflame/happy/ui/viewmodel/FriendsViewModel.kt
package com.enflame.happy.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.enflame.happy.domain.model.Friend
import com.enflame.happy.domain.model.FriendRequest
import com.enflame.happy.domain.model.FriendRequestStatus
import com.enflame.happy.domain.model.FriendStatus
import com.enflame.happy.domain.repository.FriendsRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Filter options for the friends list.
 */
enum class FriendFilter(val label: String) {
    ALL("All"),
    ONLINE("Online"),
    IN_SESSION("In Session")
}

/**
 * UI state for the friends list screen.
 *
 * @property isLoading Whether the initial load is in progress.
 * @property hasLoaded Whether the initial load has completed.
 * @property isRefreshing Whether a pull-to-refresh is in progress.
 * @property errorMessage The current error message, if any.
 * @property successMessage A transient success message to display.
 * @property searchQuery The current search query for filtering friends.
 * @property filter The current status filter.
 * @property isSendingRequest Whether a friend request is being sent.
 * @property onlineFriendCount Number of friends currently online.
 * @property pendingRequestCount Number of pending incoming friend requests.
 */
data class FriendsUiState(
    val isLoading: Boolean = false,
    val hasLoaded: Boolean = false,
    val isRefreshing: Boolean = false,
    val errorMessage: String? = null,
    val successMessage: String? = null,
    val searchQuery: String = "",
    val filter: FriendFilter = FriendFilter.ALL,
    val isSendingRequest: Boolean = false,
    val onlineFriendCount: Int = 0,
    val pendingRequestCount: Int = 0
)

/**
 * ViewModel for the Friends screens.
 *
 * Manages friends list loading, filtering, searching, friend request operations,
 * and pull-to-refresh. Observes friends and requests from the repository as
 * reactive Flows and combines with UI state for filtered results.
 */
@HiltViewModel
class FriendsViewModel @Inject constructor(
    private val friendsRepository: FriendsRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(FriendsUiState())
    val uiState: StateFlow<FriendsUiState> = _uiState.asStateFlow()

    /**
     * Raw friends from the repository, sorted by online status then name.
     */
    private val allFriends: StateFlow<List<Friend>> = friendsRepository.getFriends()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    /**
     * Friend requests from the repository.
     */
    val friendRequests: StateFlow<List<FriendRequest>> = friendsRepository.getFriendRequests()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    /**
     * Filtered and searched friends for display.
     * Combines the raw friends list with the current filter and search query.
     */
    val filteredFriends: StateFlow<List<Friend>> = combine(
        allFriends,
        _uiState
    ) { friends, state ->
        var result = friends

        // Apply status filter
        result = when (state.filter) {
            FriendFilter.ALL -> result
            FriendFilter.ONLINE -> result.filter { it.isOnline }
            FriendFilter.IN_SESSION -> result.filter { it.status == FriendStatus.IN_SESSION }
        }

        // Apply search filter
        if (state.searchQuery.isNotBlank()) {
            val query = state.searchQuery.lowercase()
            result = result.filter { friend ->
                friend.displayName.lowercase().contains(query) ||
                    friend.username.lowercase().contains(query)
            }
        }

        // Sort: online friends first (by status priority), then alphabetically
        result.sortedWith(
            compareBy<Friend> { statusSortOrder(it.status) }
                .thenBy { it.displayName.lowercase() }
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    init {
        loadFriends()
    }

    /**
     * Initial load of friends and requests from the repository.
     */
    fun loadFriends() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            try {
                friendsRepository.refreshFriends()
                friendsRepository.refreshFriendRequests()
                updateCounts()
                _uiState.update { it.copy(isLoading = false, hasLoaded = true) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        hasLoaded = true,
                        errorMessage = e.message ?: "Failed to load friends"
                    )
                }
            }
        }
    }

    /**
     * Refresh friends and requests via pull-to-refresh.
     */
    fun refreshFriends() {
        viewModelScope.launch {
            _uiState.update { it.copy(isRefreshing = true, errorMessage = null) }
            try {
                friendsRepository.refreshFriends()
                friendsRepository.refreshFriendRequests()
                updateCounts()
                _uiState.update { it.copy(isRefreshing = false) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isRefreshing = false,
                        errorMessage = e.message ?: "Failed to refresh friends"
                    )
                }
            }
        }
    }

    /**
     * Send a friend request to a user by username.
     *
     * @param username The target user's username.
     * @param message Optional message to include.
     */
    fun sendFriendRequest(username: String, message: String? = null) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSendingRequest = true, errorMessage = null) }
            try {
                friendsRepository.sendFriendRequest(username, message)
                updateCounts()
                _uiState.update {
                    it.copy(
                        isSendingRequest = false,
                        successMessage = "Friend request sent to $username"
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSendingRequest = false,
                        errorMessage = e.message ?: "Failed to send friend request"
                    )
                }
            }
        }
    }

    /**
     * Accept a pending friend request.
     *
     * @param requestId The unique identifier of the friend request.
     */
    fun acceptFriendRequest(requestId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(errorMessage = null) }
            try {
                friendsRepository.acceptFriendRequest(requestId)
                updateCounts()
                _uiState.update {
                    it.copy(successMessage = "Friend request accepted")
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        errorMessage = e.message ?: "Failed to accept friend request"
                    )
                }
            }
        }
    }

    /**
     * Decline a pending friend request.
     *
     * @param requestId The unique identifier of the friend request.
     */
    fun declineFriendRequest(requestId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(errorMessage = null) }
            try {
                friendsRepository.declineFriendRequest(requestId)
                updateCounts()
                _uiState.update {
                    it.copy(successMessage = "Friend request declined")
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        errorMessage = e.message ?: "Failed to decline friend request"
                    )
                }
            }
        }
    }

    /**
     * Remove a friend connection.
     *
     * @param friendId The unique identifier of the friend to remove.
     */
    fun removeFriend(friendId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(errorMessage = null) }
            try {
                friendsRepository.removeFriend(friendId)
                updateCounts()
                _uiState.update {
                    it.copy(successMessage = "Friend removed")
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        errorMessage = e.message ?: "Failed to remove friend"
                    )
                }
            }
        }
    }

    /**
     * Update the search query for filtering friends.
     */
    fun onSearchQueryChanged(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
    }

    /**
     * Update the status filter.
     */
    fun onFilterChanged(filter: FriendFilter) {
        _uiState.update { it.copy(filter = filter) }
    }

    /**
     * Dismiss the current error message.
     */
    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    /**
     * Dismiss the current success message.
     */
    fun dismissSuccess() {
        _uiState.update { it.copy(successMessage = null) }
    }

    /**
     * Find a friend by their ID.
     *
     * @param friendId The unique identifier of the friend.
     * @return The friend if found, null otherwise.
     */
    fun findFriend(friendId: String): Friend? {
        return allFriends.value.find { it.id == friendId }
    }

    /**
     * Update the online and pending request counts in the UI state.
     */
    private fun updateCounts() {
        val onlineCount = allFriends.value.count { it.isOnline }
        val pendingCount = friendRequests.value.count {
            it.status == FriendRequestStatus.PENDING
        }
        _uiState.update {
            it.copy(
                onlineFriendCount = onlineCount,
                pendingRequestCount = pendingCount
            )
        }
    }

    /**
     * Sort order for friend status (online first, offline last).
     */
    private fun statusSortOrder(status: FriendStatus): Int = when (status) {
        FriendStatus.IN_SESSION -> 0
        FriendStatus.ONLINE -> 1
        FriendStatus.AWAY -> 2
        FriendStatus.OFFLINE -> 3
    }
}
